File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/BetContextFactory.java
package cz.fb.manaus.reactor.betting;

import cz.fb.manaus.core.model.AccountMoney;
import cz.fb.manaus.core.model.MarketSnapshot;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.reactor.charge.ChargeGrowthForecaster;
import cz.fb.manaus.reactor.price.Fairness;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.OptionalDouble;
import java.util.Set;

@Component
public class BetContextFactory {

    @Autowired
    private ChargeGrowthForecaster forecaster;

    public BetContext create(Side side,
                             long selectionId,
                             MarketSnapshot snapshot,
                             Fairness fairness,
                             Optional<AccountMoney> accountMoney,
                             Set<String> categoryBlacklist) {
        OptionalDouble forecast = forecaster.getForecast(selectionId, side, snapshot, fairness);
        return new BetContext(side, selectionId, accountMoney, forecast, snapshot, fairness,
                categoryBlacklist);
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/BetContextFactory.kt
package cz.fb.manaus.reactor.betting

import cz.fb.manaus.core.model.Account
import cz.fb.manaus.core.model.MarketSnapshot
import cz.fb.manaus.core.model.Side
import cz.fb.manaus.reactor.charge.ChargeGrowthForecaster
import cz.fb.manaus.reactor.price.Fairness
import org.springframework.stereotype.Component

@Component
class BetContextFactory(private val forecaster: ChargeGrowthForecaster) {

    fun create(side: Side,
               selectionId: Long,
               snapshot: MarketSnapshot,
               fairness: Fairness,
               account: Account): BetContext {
        val (marketPrices, market, _, coverage, tradedVolume) = snapshot
        val commission = account.provider.commission
        val forecast = forecaster.getForecast(selectionId, side, snapshot, fairness, commission)
        val metrics = Metrics(
                chargeGrowthForecast = forecast,
                fairness = fairness,
                actualTradedVolume = tradedVolume?.get(selectionId)
        )
        return BetContext(
                market = market,
                side = side,
                selectionId = selectionId,
                marketPrices = marketPrices,
                account = account,
                coverage = coverage,
                metrics = metrics
        )
    }
}
